server/service/src/invoice_line/stock_in_line/insert/generate.rs

```diff
@@ -1,41 +1,58 @@
 use crate::{
+    barcode::{self, BarcodeInput},
     invoice::common::{calculate_total_after_tax, generate_invoice_user_id_update},
     invoice_line::{
-        convert_invoice_line_to_single_pack, convert_stock_line_to_single_pack, generate_batch,
+        convert_invoice_line_to_single_pack, convert_stock_line_to_single_pack,
+        stock_in_line::{generate_batch, StockInType, StockLineInput},
     },
     store_preference::get_store_preferences,
     u32_to_i32,
 };
 use repository::{
-    InvoiceLineRow, InvoiceLineRowType, InvoiceRow, InvoiceRowStatus, ItemRow, RepositoryError,
-    StockLineRow, StorageConnection,
+    BarcodeRow, InvoiceLineRow, InvoiceLineRowType, InvoiceRow, InvoiceRowStatus, ItemRow,
+    RepositoryError, StockLineRow, StorageConnection,
 };
 
 use super::InsertStockInLine;
 
+pub struct GenerateResult {
+    pub invoice: Option<InvoiceRow>,
+    pub invoice_line: InvoiceLineRow,
+    pub stock_line: Option<StockLineRow>,
+    pub barcode: Option<BarcodeRow>,
+}
+
 pub fn generate(
     connection: &StorageConnection,
     user_id: &str,
     input: InsertStockInLine,
     item_row: ItemRow,
     existing_invoice_row: InvoiceRow,
-) -> Result<(Option<InvoiceRow>, InvoiceLineRow, Option<StockLineRow>), RepositoryError> {
+) -> Result<GenerateResult, RepositoryError> {
     let store_preferences = get_store_preferences(connection, &existing_invoice_row.store_id)?;
 
-    let new_line = generate_line(input, item_row, existing_invoice_row.clone());
+    let new_line = generate_line(input.clone(), item_row, existing_invoice_row.clone());
 
     let mut new_line = match store_preferences.pack_to_one {
         true => convert_invoice_line_to_single_pack(new_line),
         false => new_line,
     };
 
-    let new_batch_option = if existing_invoice_row.status != InvoiceRowStatus::New {
+    let barcode_option = generate_barcode(&input, connection)?;
+
+    let new_batch_option = if should_upsert_batch(&input.r#type, &existing_invoice_row) {
         let new_batch = generate_batch(
```

Comment:
This would be should **insert** batch right ?

Reply:
I think use cases at the moment is insert only, but for other inventory adjustments, you could be creating a new `StockIn` invoice line that updates existing stock line...